include/private/soci-compiler.h

```diff
@@ -36,4 +36,40 @@
 #   define SOCI_GCC_WARNING_RESTORE(x)
 #endif
 
+// CHECK_CXX_STD(version) evaluates to 1 if the C++ version is at least the
+// version specified.
+#if defined(_MSVC_LANG)
+#    define SOCI_CHECK_CXX_STD(ver) (_MSVC_LANG >= (ver))
+#elif defined(__cplusplus)
+#    define SOCI_CHECK_CXX_STD(ver) (__cplusplus >= (ver))
+#else
+#    define SOCI_CHECK_CXX_STD(ver) 0
+#endif
+
+// HAS_CLANG_FEATURE(feature) evaluates to 1 if clang is available and the
+// provided feature is available.
+#ifndef __has_feature
+#   define SOCI_HAS_CLANG_FEATURE(x) 0
+#else
+#   define SOCI_HAS_CLANG_FEATURE(x) __has_feature(x)
+#endif
+
+// FALLTHROUGH macro defines a cross-platform/version to mark fallthrough
+// behavior in switch statement.
+#if SOCI_CHECK_CXX_STD(201703L)
+#    define SOCI_FALLTHROUGH [[fallthrough]]
+#elif defined(__has_warning) && SOCI_HAS_CLANG_FEATURE(cxx_attributes)
```

Comment:
Is `__has_warning` check really needed? I.e. can't we just do

```suggestion
#elif SOCI_HAS_CLANG_FEATURE(cxx_attributes)
```